Branch-SDK/Branch-SDK/BNCAppleSearchAds.m

```diff
@@ -0,0 +1,200 @@
+- (void)checkAppleSearchAdsSaveTo:(BNCPreferenceHelper *)preferenceHelper installDate:(NSDate *)installDate completion:(void (^_Nullable)(void))completion {
+    if (!self.enableAppleSearchAdsCheck) {
+        if (completion) {
+            completion();
+        }
+        return;
+    }
+    
+    if ([self isAppleSearchAdSavedToDictionary:preferenceHelper.appleSearchAdDetails]) {
+        if (completion) {
+            completion();
+        }
+        return;
+    }
+    
+    if (![self isDateWithinWindow:installDate]) {
+        if (completion) {
+            completion();
+        }
+        return;
+    }
+    
+    if (![self isAdClientAvailable]) {
+        if (completion) {
+            completion();
+        }
+        return;
+    }
+    
+    [self requestAttributionWithCompletion:^(NSDictionary * _Nullable attributionDetails, NSError * _Nullable error, NSTimeInterval elapsedSeconds) {
+        // BNCPreferenceHelper should be responsible for correctly storing and resetting this
+        @synchronized (preferenceHelper) {
+            if (attributionDetails.count > 0 && !error) {
+                [preferenceHelper addInstrumentationDictionaryKey:@"apple_search_ad" value:[[NSNumber numberWithInteger:elapsedSeconds*1000] stringValue]];
+            }
+            if (!error) {
+                if (attributionDetails == nil) {
+                    attributionDetails = @{};
+                }
+                if (preferenceHelper.appleSearchAdDetails == nil) {
+                    preferenceHelper.appleSearchAdDetails = @{};
+                }
+                if (![preferenceHelper.appleSearchAdDetails isEqualToDictionary:attributionDetails]) {
+                    preferenceHelper.appleSearchAdDetails = attributionDetails;
+                    preferenceHelper.appleSearchAdNeedsSend = YES;
+                }
+            }
+        }
+        
+        if (completion) {
+            completion();
+        }
+    }];
+}
+
+- (BOOL)isAppleSearchAdSavedToDictionary:(NSDictionary *)appleSearchAdDetails {
+    NSDictionary *tmp = [appleSearchAdDetails objectForKey:@"Version3.1"];
+    if (tmp && ([tmp isKindOfClass:NSDictionary.class] || [tmp isKindOfClass:NSMutableDictionary.class])) {
+        NSNumber *hasAppleSearchAdAttribution = [tmp objectForKey:@"iad-attribution"];
+        return [hasAppleSearchAdAttribution boolValue];
+    }
+    return NO;
+}
+
+- (BOOL)isDateWithinWindow:(NSDate *)installDate {
+    // install date should NOT be after current date
+    NSDate *now = [NSDate date];
+    if ([installDate compare:now] == NSOrderedDescending) {
+        return NO;
+    }
+    
+    // install date + 30 should be after current date
+    NSDate *installDatePlus30 = [installDate dateByAddingTimeInterval:(30.0*24.0*60.0*60.0)];
+    if ([installDatePlus30 compare:now] == NSOrderedDescending) {
+        return YES;
+    }
+    
+    return NO;
+}
+
+- (BOOL)isAdClientAvailable {
+    BOOL ADClientIsAvailable = self.adClientClass &&
```

Comment:
Just return this?

Reply:
This is going to be refactored with SDK-513.  Originally that check was used in two places, however it's a little redundant as it is now.